compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -1129,6 +1129,37 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     label_span_not_found();
                 }
 
+                if let SelfSource::MethodCall(expr) = source
+                    && let ty::Adt(kind, substs) = actual.kind()
+                    && tcx.get_diagnostic_item(sym::Result) == Some(kind.did())
+                {
+                    let call_expr =
+                        self.tcx.hir().expect_expr(self.tcx.hir().get_parent_node(expr.hir_id));
+                    let ok_ty = substs.type_at(0);
+                    if let Ok(pick) =
+                        self.lookup_probe(span, item_name, ok_ty, call_expr, ProbeScope::AllTraits)
+                    {
+                        err.span_note(
+                            tcx.def_span(pick.item.def_id),
+                            &format!("the method `{item_name}` exists on the type `{ok_ty}`"),
+                        );
+                        if let Some(ret_ty) = self
+                            .ret_coercion
+                            .as_ref()
+                            .map(|c| self.resolve_vars_if_possible(c.borrow().expected_ty()))
+                            && let ty::Adt(kind, _) = ret_ty.kind()
+                            && tcx.get_diagnostic_item(sym::Result) == Some(kind.did())
+                        {
+                            err.span_suggestion_verbose(
+                                expr.span.shrink_to_hi(),
+                                format!("use the `?` operator to unwrap `{actual}`"),
```

Comment:
```suggestion
                                format!("use the `?` operator to obtain the `{actual}` value if `Ok`, or return if `Err`"),
```

Not perfect, but something like this. We have to write things as if the user is encountering them for the first time, when possible :)

Reply:
“return if `Err`” doesn't really tell what is being returned. Maybe something like this would work better? 

> use the `?` operator to extract the `{actual}` value, propagating any errors to the caller